### PR TITLE
IdToken & BAM Lookup Uses Faster Cache API to Avoid Extra Decryption

### PR DESCRIPTION
**Pull request by iambmelt**
Moves idtoken lookup over to newer lookup api to avoid hitting cache multiple times (1x per tenant) when an account exists in multiple tenants

Builds on:
- [New cache API to avoid repeat decryption](https://github.com/AzureAD/microsoft-authentication-library-common-for-android/pull/1271)
- [Test fixes](https://github.com/AzureAD/microsoft-authentication-library-common-for-android/pull/1229)

### CODE REVIEW

**Review by rpdome — APPROVED** [review]

**Review by tanmaymanolkar1 — APPROVED** [review]

File: common/src/main/java/com/microsoft/identity/common/internal/cache/MsalOAuth2TokenCache.java
Comment (AdamBJohnsonx) [inline common/src/main/java/com/microsoft/identity/common/internal/cache/MsalOAuth2TokenCache.java:902]: So, thing is, the only time that it should be decrypting these values is when there is more than 256 entries in the cache.  We could tune that value.

**Review by AdamBJohnsonx — APPROVED** [review]

File: common/src/main/java/com/microsoft/identity/common/internal/cache/MsalOAuth2TokenCache.java
Comment (iambmelt) [reply common/src/main/java/com/microsoft/identity/common/internal/cache/MsalOAuth2TokenCache.java:902]: ACK - I'll go ahead and merge this change in as-is. I would love to get take your input very soon on identifying and weighing bang-for-buck opportunities to squeeze even more performance out of this layer. Work I am happy to do and think it will benefit from more eyes / creative solutions

**Comment by iambmelt** [comment]
Oddly, I cannot merge this PR despite 3 approvals....

**Comment by iambmelt** [comment]
@iamgusain Do you know why this PR cannot be merged? From looking at our code owner enforcement rules, this appears to be acceptable....

**Comment by iambmelt** [comment]
Dismissing prior reviews because I may have gone too crazy in these latest changes 😆

File: common/src/main/java/com/microsoft/identity/common/internal/cache/BrokerOAuth2TokenCache.java
Comment (AdamBJohnsonx) [inline common/src/main/java/com/microsoft/identity/common/internal/cache/BrokerOAuth2TokenCache.java:1207]: We should NOT do this until we can measure it.  If there aren't enough execution units to manage it, it will be slower than not.

Comment (iambmelt) [reply common/src/main/java/com/microsoft/identity/common/internal/cache/BrokerOAuth2TokenCache.java:1207]: ACK - holding....

Comment (shahzaibj) [reply common/src/main/java/com/microsoft/identity/common/internal/cache/BrokerOAuth2TokenCache.java:1207]: I believe we don't need to wait for Paul's perf work to measure this specific thing...that's more E2E perf measurement...over here I think we can just run a test with both old and new code and just capture numbers in the test and assert that the new code takes < time than old code (we can probably capture and assert on a percentage increase)

Also this is a sort of change that I think is also easy to flight...so if we have ability to do that now then we should try that too.

File: common/src/main/java/com/microsoft/identity/common/internal/cache/BrokerOAuth2TokenCache.java
Comment (shahzaibj) [inline common/src/main/java/com/microsoft/identity/common/internal/cache/BrokerOAuth2TokenCache.java:1689]: seems like an attempt to distinguish an overload....how do we disntiguish the following?

```java
 private void methodName(TypeA input)
```

```java
 private void methodName(TypeB input)
```

File: common/src/main/java/com/microsoft/identity/common/internal/cache/BrokerOAuth2TokenCache.java
Comment (shahzaibj) [inline common/src/main/java/com/microsoft/identity/common/internal/cache/BrokerOAuth2TokenCache.java:1721]:
```suggestion
     * Returns the {@link OAuth2TokenCache} to use for supplied client id and environment.
```

File: common/src/main/java/com/microsoft/identity/common/internal/cache/BrokerOAuth2TokenCache.java
Comment (AdamBJohnsonx) [reply common/src/main/java/com/microsoft/identity/common/internal/cache/BrokerOAuth2TokenCache.java:1207]: We don't really have that ability, we could set it up, but nothing is populating the flight info right now, and we're not integrate with any metrics system that I know of.  The only thing I'm cautioning on is the parallelization.  Do we really know what we're getting into here, and will it actually have the effect we want?  If we want to tackle this, - don't create new thread pools to handle it each time, don't used a cached pool, we MUST measure it, and we need to monitor it.

Comment (iambmelt) [reply common/src/main/java/com/microsoft/identity/common/internal/cache/BrokerOAuth2TokenCache.java:1204]: Reverted these changes in 6c0625b886331da1e58d55f8604965b22d81fc4d -- not going to try this until we can measure/evaluate

File: common/src/main/java/com/microsoft/identity/common/internal/cache/BrokerOAuth2TokenCache.java
Comment (iambmelt) [reply common/src/main/java/com/microsoft/identity/common/internal/cache/BrokerOAuth2TokenCache.java]: Made less ambiguous in ceea53560b6510d9d363a9fbc49c7dff5812da74

**Comment by iambmelt** [comment]
Have backed out the more questionable changes from this PR and clarified log tags... rerequesting reviews

**Review by AdamBJohnsonx — APPROVED** [review]

File: common/src/main/java/com/microsoft/identity/common/internal/cache/MsalOAuth2TokenCache.java
Comment (rpdome) [inline common/src/main/java/com/microsoft/identity/common/internal/cache/MsalOAuth2TokenCache.java:904]: nit: please add comments on the List<Credential> parameter for this overload

**Review by rpdome — APPROVED** [review]

**Review by shahzaibj — APPROVED** [review]

**Comment by iamgusain** [comment]
As you are touching the "cache" area, It needs signoff from code owners for cache..  you and Shane at the moment.
And it doesnt allow you to signoff your own PR. 

---
In reply to: [844639652](https://github.com/AzureAD/microsoft-authentication-library-common-for-android/pull/1385#issuecomment-844639652) [](http://example.com/codeflow?ancestors=844639652,844639652)

**Review by tanmaymanolkar1 — APPROVED** [review]

File: common/src/main/java/com/microsoft/identity/common/internal/cache/MsalOAuth2TokenCache.java
Comment (iambmelt) [inline common/src/main/java/com/microsoft/identity/common/internal/cache/MsalOAuth2TokenCache.java:902]:
```suggestion
        // Load all the credentials to inspect once, such that we don't need to requery the cache
        // pass these into the new getCredentialsFilteredBy overload, rather than hit disk again
        final List<Credential> allCredentials = mAccountCredentialCache.getCredentials();
```

File: common/src/main/java/com/microsoft/identity/common/internal/cache/MsalOAuth2TokenCache.java
Comment (iambmelt) [reply common/src/main/java/com/microsoft/identity/common/internal/cache/MsalOAuth2TokenCache.java:906]: 82e4ab0c9065759ed3aad0bda5b50eec3f3a2b7e

**Comment by iambmelt** [comment]
> As you are touching the "cache" area, It needs signoff from code owners for cache.. you and Shane at the moment.
> And it doesnt allow you to signoff your own PR.
> 
> In reply to: [844639652](https://github.com/AzureAD/microsoft-authentication-library-common-for-android/pull/1385#issuecomment-844639652)

@iamgusain I am going to add you as a codeowner then, because in this state I can't merge

**Review by iamgusain — APPROVED** [review]
:shipit:

File: common/src/main/java/com/microsoft/identity/common/internal/cache/BrokerOAuth2TokenCache.java
Comment (shoatman) [inline common/src/main/java/com/microsoft/identity/common/internal/cache/BrokerOAuth2TokenCache.java:1706]: Probably worth calling out somewhere that this identifier is unique and assigned at installation time.  With that said I don't think anything stops these identifiers from being recycled.

**Review by shoatman — APPROVED** [review]
Fee free to ignore my comment.

File: common/src/main/java/com/microsoft/identity/common/internal/cache/BrokerOAuth2TokenCache.java
Comment (iambmelt) [inline common/src/main/java/com/microsoft/identity/common/internal/cache/BrokerOAuth2TokenCache.java:1706]:
```suggestion
     * @param callingProcessUid The uid of the calling process. This value is unique to the calling app and is assigned during installation.
```

Comment (iambmelt) [reply common/src/main/java/com/microsoft/identity/common/internal/cache/BrokerOAuth2TokenCache.java]: b70689bf406336908d169895bc1f7408db842c96

**Pull request merged**
